File: cpp/creational/prototype/nishith/Prototype/Prototype/Player.cpp
#include "Player.h"
#include <iostream>


void Player::print() const
{
   std::cout << "Player: " << name_ << ", Health: " << health_
      << ", Damage Output: " << damage_ << std::endl;
}

Player* Player::clone() const  
{  
  return new Player(*this);  
}

Player::Player(const std::string& name, int health, int damage)
   : name_(name), health_(health), damage_(damage)
{
}

File: java/creational/prototype/uday/Player.java
/**
 * Abstract base class representing a Player in the game.
 * Defines common properties like name and health, and abstract methods
 * for cloning and printing player details.
 */
public abstract class Player {
    
    private String name;
    private int health;
  
    /**
     * Constructs a Player with given name and health.
     */
    public Player(String name, int score) {
        this.name = name;
        this.health = score;
    }

    /**
     * Copy constructor to clone a Player.
     */
    public Player(Player player) {
        this.name = player.name;
        this.health = player.health;
    }

    public String getName() {
        return name;
    }

    public int getHealth() {
        return health;
    }

    /**
     * Abstract method to create a copy of the Player.
     */
    public abstract Player clone();

    /**
     * Abstract method to print player information.
     */
    public abstract void print();
}
